[isoltest] Add gas costs to function call expectations

mijovic opened the pull request:
Fixes #10474

TODO:
- [x] Add expected gas costs, not only obtained
- [ ] Add deploy costs
- [x] Run ir gas cost only if explicitly added to expectations

chriseth commented on test/libsolidity/semanticTests/abiEncoderV1/abi_decode_static_array.sol:
```diff
@@ -12,3 +12,7 @@ contract C {
 // compileViaYul: also
 // ----
 // f(bytes): 0x20, 0xc0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6 -> 1, 2, 3, 4, 5, 6
+// gas Legacy: 30262
```
Doesn't this depend on the evm version?

mijovic replied:
We decided to go with latest EVM, Will change gas costs to match latest

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -295,6 +295,10 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
 bool SemanticTest::checkGasCostExpectation(TestFunctionCall& _test, bool _compileViaYul) const
 {
+	// TODO: check other evm versions
+	if (m_evmVersion != EVMVersion::istanbul())
```
Why not always latest?

mijovic replied:
I was running default setup for isoltest and it was istanbul for me (is it dependant on which version I have installed?).
I can do latest in any case.

cameel replied:
Is there a way to get the latest one other than `EVMVersion{}`? Something like `EVMVersion::latest()` would be more readable.

cameel replied:
@mijovic 
> (is it dependant on which version I have installed?).

I think `istanbul` is the default now and you can choose a different one using the `--evm-version` option:

```bash
soltest -- --evm-version istanbul
```
```bash
isoltest --evm-version istanbul
```

In CI we're running tests on all versions.

mijovic replied:
Changed to latest

cameel replied:
I asked about `berlin` on gitter and here latest should be `istanbul`. It's the latest stable version. So this should be set to `EVMVersion{}`, not `EVMVersion::latest()`.

In that case I wonder if it made sense to add the `latest()` helper after all. It looks like the one you would want by default but that's misleading. I originally assumed that latest and default would be pretty much the same thing for us but they're not.

mijovic replied:
Removed latest helper and run only on istanbul now

cameel replied:
It's `istanbul` now which is good but I maybe I wasn't clear that I think you should actually use `EVMVersion{}` (i.e. using the default constructor), not `EVMVersion::istanbul()`. This way we won't have to find and update this when we switch the default in that class. [That's what's used for the `--evm-version` option](https://github.com/ethereum/solidity/blob/v0.8.1/test/Common.cpp#L182-L193).

chriseth replied:
This still needs to be addressed, right?

chriseth commented:
I'm not sure it's a good idea to add the gas costs always for all test.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `1be489917b2431f2d7e93c5319f64529e628325c`:
```
Error: Trailing whitespace found:
test/libsolidity/semanticTests/ecrecover/ecrecover_abiV2.sol:13:// 0x73b1693892219d736caba55bdb67216e485557ea6b6af75f37096c9aa6a5a75f, 

```
Please check that your changes are working as intended.

axic commented on test/libsolidity/semanticTests/abiEncoderV1/abi_decode_dynamic_array.sol:
```diff
@@ -8,3 +8,7 @@ contract C {
 // f(bytes): 0x20, 0xc0, 0x20, 0x4, 0x3, 0x4, 0x5, 0x6 -> 0x20, 0x4, 0x3, 0x4, 0x5, 0x6
+// gas Legacy: 25792
+// gas LegacyOptimized: 23842
+// gas Yul: 26117
```
In #10779 there seems to be a convergence to wards "legacy" and "ir" as the two options. Perhaps we should use those terms here too.

I am fine as is, and we can do another big round of renaming as part of the PR changing `compileViaYul`.

cameel replied:
Naming came up in https://github.com/ethereum/solidity/pull/10356#discussion_r565514043 too. Do you think we should use IR instead of Yul there?

mijovic replied:
Will rename to "ir"

chriseth commented:
We should also add deploy gas costs, or maybe better: code size

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -271,9 +277,37 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
+		return true;
+
+	string setting =
+		(_compileViaYul ? string{"Yul"} : "Legacy") +
```
```suggestion
		(_compileViaYul ? "Yul"s : "Legacy"s) +
```

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -271,9 +277,37 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
+	_test.setGasCost(setting, m_gasUsed);
+	return
+		_test.call().expectations.gasUsed.count(setting) > 0 &&
+		m_gasUsed == _test.call().expectations.gasUsed.at(setting);
```
Maybe we should add some tolerance here so that we don't have to update tests for minuscule differences?

axic replied:
This was actually discussed on gitter in December and the consensus seemed to be strict values without tolerance.

cameel replied:
Oh, ok. I remember some discussion about it on the channel. In that case it's fine.

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -321,7 +355,7 @@ void SemanticTest::printSource(ostream& _stream, string const& _linePrefix, bool
 {
 	for (auto const& test: m_tests)
-		_stream << test.format("", true, false) << endl;
+		_stream << test.format("", !m_gasCostFailure, false);
```
Being able to see the names of these boolean parameters would be very helpful when reviewing:
```suggestion
		_stream << test.format("", /* _renderResult = */ !m_gasCostFailure, /* _highlight = */ false);
```

mijovic replied:
Added. I do not have this habit as my IDE (CLion) adds annotations for me

cameel replied:
Yeah, it's not that much of a problem when I'm using an editor and can jump to definition but I can't do that on github :)

And it's not something I need all the time but in this function in particular the parameters were a complete mystery to me until I checked :)

mijovic replied:
Completely agree with you on this :)

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -321,7 +355,7 @@ void SemanticTest::printSource(ostream& _stream, string const& _linePrefix, bool
 void SemanticTest::printUpdatedExpectations(ostream& _stream, string const&) const
 {
 	for (auto const& test: m_tests)
-		_stream << test.format("", true, false) << endl;
+		_stream << test.format("", !m_gasCostFailure, false);
```
The docstring for the `format()` function says:
```c++
	/// Formats this function call test and applies the format that was detected during parsing.
	/// If _renderResult is false, the expected result of the call will be used, if it's true
	/// the actual result is used.
	/// If _highlight is false, it's formatted without colorized highlighting. If it's true, AnsiColorized is
	/// used to apply a colorized highlighting.
	/// If test expectations do not match, the contract ABI is consulted in order to get the
	/// right encoding for returned bytes, based on the parsed return types.
	/// Reports warnings and errors to the error reporter.
```

So if I understand this correctly, your change means to use the expected gas cost if gas check failed and actual gas cost if it passed. But if it passed, the actual cost must have been the same as the expectation. So it's always the expected cost and you could just set it to `false`.

This would actually make sense if there was a tolerance on the cost comparison but you're doing an exact comparison instead.

BTW, this is the only place outside of `runTest()` that uses `m_gasCostFailure` so if it turns out that you do change it to `false`, then I think the variable should be converted to a local variable.

mijovic replied:
I wanted to call with false in case of gas cost failure, and true in case of regular expectations failure. This is used to print error messages and update expectations accordingly.. 
There was a wierd issue when function call expects failure, and if I pass `true` here in case of only gas failure it changes expectations in a way that translates strings to raw bytes. So, I wanted to make sure gas costs doesn't touch other expectations with this flag...

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -271,9 +277,37 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
 	return TestResult::Success;
 }
 
+bool SemanticTest::checkGasCostExpectation(TestFunctionCall& _test, bool _compileViaYul) const
```
`_test.call()` is `const` so I think you should be able to make the `_test` parameter const too.

Otherwise it should be called `io_test` according to our convention (which is ugly but does bring the attention to the fact that it's being modified).

```suggestion
bool SemanticTest::checkGasCostExpectation(TestFunctionCall const& _test, bool _compileViaYul) const
```

mijovic replied:
There is a call to 1setGasCost1 in that function, that is why there is no `const`
`_test.setGasCost(setting, m_gasUsed);`

mijovic replied:
Renamed to `io_test` to match convention

mijovic commented:
I'll change PR now, so that gas is checked only for these tests where expectation is present and will try filtering out tests where we want to run at the beginning.

mijovic committed:
Running gas tests only on tests having expectations.

mijovic committed:
Small style fix for SemanticTest constructor.

mijovic commented:
@cameel Updated
Gas tests will be run only when expectation is present.
I removed gas expectations from all the tests that have less than 100K gas usage (still 600+ tests are affected)
Also, updated to use only latest version of evm for gas usage

cameel commented on test/libsolidity/semanticTests/abiEncoderV2/cleanup/dynamic_array.sol:
```diff
@@ -14,8 +14,8 @@ contract C {
 }
 // ====
-// compileViaYul: also
 // EVMVersion: >homestead
+// compileViaYul: also
```
This PR would be shorter and a bit less intimidating if you could remove changes that now don't add gas expectations but only move this line.

mijovic replied:
I have implemented enforcing gas tests with some threshold value and used that to activate tests that have usage greater than 100000.

I think it should be fine now

mijovic committed:
Implementing enforcing gas expectations in isoltest.

Co-authored-by: Daniel Kirchner <daniel@ekpyron.org>

cameel commented on liblangutil/EVMVersion.h:
```diff
@@ -51,6 +51,7 @@ class EVMVersion:
 	static EVMVersion petersburg() { return {Version::Petersburg}; }
 	static EVMVersion istanbul() { return {Version::Istanbul}; }
 	static EVMVersion berlin() { return {Version::Berlin}; }
+	static EVMVersion latest() { return {Version::Berlin}; }
```
Nice.

cameel replied:
By the way, I just realized that we already have `berlin` there. I thought we didn't because we don't run soltest on that in CI (https://github.com/ethereum/solidity/blob/v0.8.1/.circleci/soltest_all.sh#L31).

@chriseth Should we add a soltest run for `berlin`?

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -271,9 +286,44 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
 
+bool SemanticTest::checkGasCostExpectation(TestFunctionCall& io_test, bool _compileViaYul) const
+{
+	// TODO: check other evm versions
+	if (m_evmVersion != EVMVersion::latest())
+		return true;
```
I think you can remove this TODO now. I think that the decision on the channel was to run checks only on the latest version?

Also, maybe it would be better if `soltest`/`isoltest` returned an error when you use `--enforce-gas-test` together with `--evm-version` and do not specify the latest EVM? Then you could make this an assert. Currently the requirements will just be silently ignored.

mijovic replied:
Added assertThrow in validate for `CommonOptions` it throws properly in soltest. Also, in isoltest I added it to be default true in case we are running on latest evm. So, this should be fine now

cameel commented on test/Common.cpp:
```diff
@@ -103,6 +103,8 @@ CommonOptions::CommonOptions(std::string _caption):
 		("optimize", po::bool_switch(&optimize), "enables optimization")
 		("enforce-via-yul", po::bool_switch(&enforceViaYul), "Enforce compiling all tests via yul to see if additional tests can be activated.")
+		("enforce-gas-test", po::bool_switch(&enforceGasTest), "Enforce running gas tests on every test and adding expectations.")
+		("enforce-gas-test-min-value", po::value(&enforceGasTestMinValue), "Threshold value when enforcing gas tests.")
```
Why not `--enforce-gas-cost`? That would match the variable name and we already have separate gas tests so it's not clear which gas tests this refers to.

```suggestion
		("enforce-gas-cost", po::bool_switch(&enforceGasTest), "Enforce gas cost expectations in semantic tests.")
		("enforce-gas-cost-min-value", po::value(&enforceGasTestMinValue), "Threshold value when enforcing gas cost expectations.")
```

mijovic commented on test/libsolidity/semanticTests/abiEncoderV2/abi_encode_v2.sol:
```diff
@@ -48,8 +48,8 @@ contract C {
-// f2() -> 0x20, 0xa0, 0x1, 0x60, 0x2, 0x3, "abc"
-// f3() -> 0x20, 0xa0, 0x1, 0x60, 0x2, 0x3, "abc"
-// f4() -> 0x20, 0x160, 0x1, 0x80, 0xc0, 0x2, 0x3, "abc", 0x7, 0x40, 0x2, 0x2, 0x3
+// f0() -> 0x20, 0x0// f1() -> 0x20, 0x40, 0x1, 0x2// f2() -> 0x20, 0xa0, 0x1, 0x60, 0x2, 0x3, "abc"// f3() -> 0x20, 0xa0, 0x1, 0x60, 0x2, 0x3, "abc"// f4() -> 0x20, 0x160, 0x1, 0x80, 0xc0, 0x2, 0x3, "abc", 0x7, 0x40, 0x2, 0x2, 0x3
```
Need to fix this

cameel commented on test/TestCase.h:
```diff
@@ -41,6 +41,8 @@ class TestCase
 		std::vector<boost::filesystem::path> vmPaths;
 		bool enforceCompileViaYul;
+		bool enforceGasCost = false;
+		u256 enforceGasCostMinValue = 100000;
```
I see that `enforceCompileViaYul` does not have a default here and has one in `CommonOptions`. I guess it's being set to the value from `CommonOptions` somewhere down the line?

If so, could you use the same mechanism for your options? It would be best to have defaults hard-coded in as few places as possible.

cameel commented on test/libsolidity/SemanticTest.h:
```diff
@@ -70,6 +88,10 @@ class SemanticTest: public SolidityExecutionFramework, public EVMVersionRestrict
+
+	bool m_gasCostFailure = false;
+	bool m_enforceGasCost = false;
+	u256 m_enforceGasCostMinValue = 100000;
```
These don't need defaults here. You already have defaults for these parameters in the only available constructor so the defaults set here will never actually be used.

cameel commented on test/libsolidity/SemanticTest.h:
```diff
@@ -40,9 +40,26 @@ class SemanticTest: public SolidityExecutionFramework, public EVMVersionRestrict
+				_options.enforceCompileViaYul,
+				_options.enforceGasCost,
+				_options.enforceGasCostMinValue
+			);
```
```suggestion
		return std::make_unique<SemanticTest>(
			_options.filename,
			_options.evmVersion,
			_options.vmPaths,
			_options.enforceCompileViaYul,
			_options.enforceGasCost,
			_options.enforceGasCostMinValue
		);
```

cameel commented on test/libsolidity/semanticTests/abiEncoderV2/calldata_array.sol:
```diff
@@ -15,9 +15,15 @@ contract C {
 // EVMVersion: >homestead
+// compileViaYul: also
 // ----
 // f(uint256[][1]): 32, 32, 0 -> true
+// gas ir: 105308
 // f(uint256[][1]): 32, 32, 1, 42 -> true
+// gas ir: 139370
 // f(uint256[][1]): 32, 32, 8, 421, 422, 423, 424, 425, 426, 427, 428 -> true
+// gas ir: 377825
+// gas legacy: 144300
+// gas legacyOptimized: 124231
+// gas irOptimized: 227663
```
Why is isoltest inserting only `gas ir` here? Is the 100k check done separately for each of these parameters?

mijovic replied:
Yes it is actually. Do you think it should do it differently?

cameel replied:
I'm not sure. Both are valid ways to do this and to be honest I'm not sure which one is better. I commented mostly because I expected them all to be printed.

I think we need more opinions here.

mijovic replied:
Let's wait for more feedback before updating then :-)

chriseth replied:
I think it makes sense as it is now. In addition to that: I would leave out `gas ir` always except if it is given explicitly. Codegen via IR is meant to be run with the optimizer, so the number should always be ignored unless maybe if we have a test that wants to check if a certain construct is optimized away.

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -212,7 +224,10 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
 				);
 			}
 
-			bool outputMismatch = (output != test.call().expectations.rawBytes());
+			bool outputMismatch = output != test.call().expectations.rawBytes();
```
I'd keep the parenthesis. I usually add them myself in expressions containing both assignment and boolean or comparison operators for clarity.

cameel commented on test/libsolidity/semanticTests/array/pop/array_pop_array_transition.sol:
```diff
@@ -25,4 +25,8 @@ contract c {
 // compileViaYul: also
 // ----
 // test() -> 1, 2, 3
+// gas ir: 2556862
+// gas legacy: 2416722
+// gas legacyOptimized: 2405432
+// gas irOptimized: 2556862
```
This one is really expensive...

cameel commented on test/libsolidity/semanticTests/freeFunctions/free_namesake_contract_function.sol:
```diff
@@ -5,7 +5,11 @@ contract C {
 // compileToEwasm: also
+// compileViaYul: also
 // ----
 // f() -> FAILURE
+// gas ir: 100000000
+// gas legacy: 100000000
+// gas legacyOptimized: 100000000
+// gas irOptimized: 100000000
```
Is this one so expensive because the failure eats all remaining gas?

mijovic replied:
Looks like that

cameel commented on test/libsolidity/semanticTests/functionCall/failed_create.sol:
```diff
@@ -14,15 +14,25 @@ contract C {
 // EVMVersion: >=byzantium
+// compileViaYul: also
 // ----
 // constructor(), 20 wei
+// gas ir: 437069
+// gas legacy: 324594
+// gas legacyOptimized: 222226
+// gas irOptimized: 232551
 // f(uint256): 20 -> 1370859564726510389319704988634906228201275401179
 // x() -> 1
 // f(uint256): 20 -> FAILURE
 // x() -> 1
 // stack(uint256): 1023 -> FAILURE
+// gas ir: 1162959
+// gas legacy: 981671
+// gas legacyOptimized: 831459
+// gas irOptimized: 835314
 // x() -> 1
 // stack(uint256): 10 -> 693016686122178122849713379390321835634789309880
+// gas ir: 106194
+// gas legacy: 104853
 // x() -> 2
```
Another expensive one.

mijovic replied:
It runs recursion, I think that is why it makes it so expensive

cameel commented on test/libsolidity/semanticTests/salted_create/salted_create.sol:
```diff
@@ -17,8 +17,15 @@ contract A {
 // EVMVersion: >=constantinople
+// compileViaYul: also
 // ----
 // different_salt() -> true
+// gas ir: 107186
+// gas legacy: 125934
+// gas legacyOptimized: 125869
 // same_salt() -> true
+// gas ir: 98439503
+// gas legacy: 98439401
+// gas legacyOptimized: 98439300
```
The difference is surprisingly big here. Does try/catch in `same_salt()` really cost that much or is it due to a different salt?

mijovic replied:
Not sure, but @chriseth should probably know

cameel commented on test/libsolidity/semanticTests/various/staticcall_for_view_and_pure.sol:
```diff
@@ -32,9 +32,20 @@ contract D {
 // EVMVersion: >=byzantium
+// compileViaYul: also
 // ----
 // f() -> 0x1 # This should work, next should throw #
+// gas ir: 186425
+// gas legacy: 121180
+// gas legacyOptimized: 111588
 // fview() -> FAILURE
+// gas ir: 98440114
+// gas legacy: 98439107
+// gas legacyOptimized: 98438963
+// gas irOptimized: 98438674
 // fpure() -> FAILURE
+// gas ir: 98440114
+// gas legacy: 98439107
+// gas legacyOptimized: 98438963
+// gas irOptimized: 98438674
```
Expensive.

cameel commented on test/libsolidity/semanticTests/viaYul/array_storage_index_access.sol:
```diff
@@ -18,11 +18,42 @@ contract C {
 // test_indices(uint256): 129 ->
+// gas ir: 3797603
+// gas legacy: 3340105
+// gas legacyOptimized: 3280773
+// gas irOptimized: 3572413
 // test_indices(uint256): 5 ->
+// gas ir: 781923
+// gas legacy: 458941
+// gas legacyOptimized: 455849
+// gas irOptimized: 684693
 // test_indices(uint256): 10 ->
+// gas ir: 107663
 // test_indices(uint256): 15 ->
+// gas ir: 139793
+// gas irOptimized: 115342
 // test_indices(uint256): 0xFF ->
+// gas ir: 4994703
+// gas legacy: 4107867
+// gas legacyOptimized: 3991807
+// gas irOptimized: 4554277
 // test_indices(uint256): 1000 ->
+// gas ir: 23586895
+// gas legacy: 20360399
+// gas legacyOptimized: 19921344
+// gas irOptimized: 21917394
 // test_indices(uint256): 129 ->
+// gas ir: 5948697
+// gas legacy: 3472135
+// gas legacyOptimized: 3415947
+// gas irOptimized: 5135327
 // test_indices(uint256): 128 ->
+// gas ir: 852771
+// gas legacy: 556972
+// gas legacyOptimized: 508124
+// gas irOptimized: 663783
 // test_indices(uint256): 1 ->
+// gas ir: 773769
+// gas legacy: 452407
+// gas legacyOptimized: 450811
+// gas irOptimized: 680239
```
Expensive.

cameel commented on test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol:
```diff
@@ -54,6 +54,22 @@ contract C {
 // ----
 // test_zeroed_indicies(uint256): 1 ->
 // test_zeroed_indicies(uint256): 5 ->
+// gas ir: 243517
+// gas legacy: 191267
+// gas legacyOptimized: 188486
+// gas irOptimized: 222625
 // test_zeroed_indicies(uint256): 10 ->
+// gas ir: 363409
+// gas legacy: 276129
+// gas legacyOptimized: 271024
+// gas irOptimized: 327375
 // test_zeroed_indicies(uint256): 15 ->
+// gas ir: 460299
+// gas legacy: 339254
+// gas legacyOptimized: 331904
+// gas irOptimized: 409400
 // test_zeroed_indicies(uint256): 0xFF ->
+// gas ir: 10516489
+// gas legacy: 8477449
+// gas legacyOptimized: 8343774
+// gas irOptimized: 9686975
```
Expensive.

cameel commented on test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:
```diff
@@ -14,5 +14,17 @@ contract C {
 // set_get_length(uint256): 10 -> 10
 // set_get_length(uint256): 20 -> 20
 // set_get_length(uint256): 0xFF -> 0xFF
+// gas ir: 917394
+// gas legacy: 619622
+// gas legacyOptimized: 600751
+// gas irOptimized: 434473
 // set_get_length(uint256): 0xFFF -> 0xFFF
+// gas ir: 14623616
+// gas legacy: 9765519
+// gas legacyOptimized: 9461853
+// gas irOptimized: 6743235
 // set_get_length(uint256): 0xFFFFF -> FAILURE # Out-of-gas #
+// gas ir: 100000000
+// gas legacy: 100000000
+// gas legacyOptimized: 100000000
+// gas irOptimized: 100000000
```
Expensive.

cameel commented on test/libsolidity/semanticTests/viaYul/array_storage_push_empty_length_address.sol:
```diff
@@ -10,14 +10,30 @@ contract C {
 // EVMVersion: >=petersburg
+// compileViaYul: also
 // ----
 // set_get_length(uint256): 0 -> 0
 // set_get_length(uint256): 1 -> 1
 // set_get_length(uint256): 10 -> 10
 // set_get_length(uint256): 20 -> 20
 // set_get_length(uint256): 0 -> 0
+// gas ir: 144209
+// gas legacy: 107830
+// gas legacyOptimized: 107295
+// gas irOptimized: 110296
 // set_get_length(uint256): 0xFF -> 0xFF
+// gas ir: 994431
+// gas legacy: 882337
+// gas legacyOptimized: 650737
+// gas irOptimized: 702388
 // set_get_length(uint256): 0xFFF -> 0xFFF
+// gas ir: 14624613
+// gas legacy: 12945874
+// gas legacyOptimized: 9462679
+// gas irOptimized: 10238500
 // set_get_length(uint256): 0xFFFF -> FAILURE # Out-of-gas #
+// gas ir: 100000000
+// gas legacy: 100000000
+// gas legacyOptimized: 100000000
+// gas irOptimized: 100000000
```
Expensive.

cameel commented on test/libsolidity/semanticTests/viaYul/array_storage_push_pop.sol:
```diff
@@ -14,7 +14,24 @@ contract C {
 // set_get_length(uint256): 0 -> 0
 // set_get_length(uint256): 1 -> 0
 // set_get_length(uint256): 10 -> 0
+// gas ir: 121925
 // set_get_length(uint256): 20 -> 0
+// gas ir: 218885
+// gas legacy: 141922
+// gas legacyOptimized: 139741
+// gas irOptimized: 162779
 // set_get_length(uint256): 0xFF -> 0
+// gas ir: 2497445
+// gas legacy: 1524427
+// gas legacyOptimized: 1500391
+// gas irOptimized: 1792504
 // set_get_length(uint256): 0xFFF -> 0
+// gas ir: 39730097
+// gas legacy: 24115159
+// gas legacyOptimized: 23734003
+// gas irOptimized: 28422916
 // set_get_length(uint256): 0xFFFF -> FAILURE # Out-of-gas #
+// gas ir: 100000000
+// gas legacy: 100000000
+// gas legacyOptimized: 100000000
+// gas irOptimized: 100000000
```
Expensive.

axic replied:
That is the tx gas limit, failing transactions use up the tx gas limit :) That's why we switched to `revert` where possible.

cameel commented on test/libsolidity/util/TestFunctionCall.h:
```diff
@@ -124,6 +128,8 @@ class TestFunctionCall
 	/// Result of the actual call been made.
 	bytes m_rawBytes = bytes{};
+	/// Actual gas cost for the type of the run
+	std::pair<std::string, u256> m_gasCost;
```
```suggestion
	std::tuple<std::string, u256> m_gasCost;
```

cameel commented on test/libsolidity/util/TestFunctionCall.h:
```diff
@@ -116,6 +117,9 @@ class TestFunctionCall
 		std::string const& _linePrefix = ""
 	) const;
 
+	/// Formats gas used expectations one per line
```
```suggestion
	/// Formats gas usage expectations one per line
```

cameel commented on test/tools/IsolTestOptions.cpp:
```diff
@@ -76,6 +76,8 @@ bool IsolTestOptions::parse(int _argc, char const* const* _argv)
 		return false;
 	}
 	enforceViaYul = true;
+	enforceGasTest = true;
+	enforceGasTestMinValue = 100000;
```
Maybe we could have some global constants for these defaults? They get repeated a lot.

mijovic replied:
True that, plus maybe will remove from few places as well

mijovic replied:
After removing it is present only in one place, I think it is fine now

cameel commented on test/libsolidity/util/TestFileParser.cpp:
```diff
@@ -77,76 +75,99 @@ vector<solidity::frontend::test::FunctionCall> TestFileParser::parseFunctionCall
+
+						string runType = m_scanner.currentLiteral();
+						if (set<string>{"ir", "irOptimized", "legacy", "legacyOptimized"}.count(runType) > 0)
+						{
+							m_scanner.scanNextToken();
+							expect(Token::Colon);
+							calls.back().expectations.gasUsed[runType] = u256(parseDecimalNumber());
+						}
 						else
-							throw TestParserError("Expected \"empty\" or \"nonempty\".");
-						call.kind = FunctionCall::Kind::Storage;
-						m_scanner.scanNextToken();
+							throw TestParserError(
+								"Expected \"ir\", \"irOptimized\", \"legacy\", or \"legacyOptimized\"."
+							);
+
 					}
 					else
 					{
-						bool lowLevelCall = false;
-						tie(call.signature, lowLevelCall) = parseFunctionSignature();
-						if (lowLevelCall)
-							call.kind = FunctionCall::Kind::LowLevel;
+						FunctionCall call;
 
```
How do you handle duplicate expectations? E.g.

```c++
// gas ir: 1
// gas legacy: 2
// gas ir: 3
```

mijovic replied:
Good point, will add error in case we have duplicate expectations. So far, I think it would just use second value and pass the test if it is correct...

cameel reviewed (commented):
Here's another batch of comments.

I also skimmed through the updated tests and commented on the ones where the cost is particularly high (>= 1 million gas) so that we can discuss them but I didn't really analyze them to check if the high cost is warranted or not. Feel free to mark any obvious ones as resolved.

mijovic commented:
I think I addressed all comments, will wait for points we need more feedback.

aarlt commented on test/libsolidity/semanticTests/array/copying/copy_byte_array_in_struct_to_storage.sol:
```diff
@@ -37,6 +37,13 @@ contract C {
+// gas legacyOptimized: 174212
+// gas irOptimized: 172892
 // g() -> 0x40, 0xc0, 0x49, 0x3132333435363738393031323334353637383930313233343536373839303120, 0x3132333435363738393031323334353637383930313233343536373839303120, 0x3132333435363738390000000000000000000000000000000000000000000000, 0x11, 0x3132333435363738393233343536373839000000000000000000000000000000
+// gas ir: 106397
```
Why is `gas irOptimized` missing here?

cameel replied:
See https://github.com/ethereum/solidity/pull/10863#discussion_r566840010

aarlt commented on test/libsolidity/semanticTests/array/copying/copy_internal_function_array_to_storage.sol:
```diff
@@ -18,8 +18,10 @@ contract C {
+// compileViaYul: also
 // ----
 // one() -> 3
+// gas legacy: 154760
```
Why is `gas ir*`  missing here? I would have expected to see `gas ir*` expectations here too, because `compileViaYul` is defined.

cameel replied:
See https://github.com/ethereum/solidity/pull/10863#discussion_r566840010

aarlt commented on test/libsolidity/semanticTests/array/pop/array_pop.sol:
```diff
@@ -15,3 +15,4 @@ contract c {
 // compileViaYul: also
 // ----
 // test() -> 1, 0
+// gas ir: 101390
```
Looks like I don't understand what exactly decides what kind of gas measurement is done: why is the legacy gas measurement not defined here too?

aarlt commented on test/libsolidity/semanticTests/functionCall/failed_create.sol:
```diff
@@ -14,15 +14,25 @@ contract C {
+// gas irOptimized: 835314
 // x() -> 1
 // stack(uint256): 10 -> 693016686122178122849713379390321835634789309880
+// gas ir: 106194
```
What exactly decides when the gas measurement for optimized code need to be done?

aarlt commented:
Maybe it would be interesting to add test cases that are doing gas measurements for all the different evm versions. E.g. setting the same tests that only differ in the defined `EVMVersion`. With this we could more easily detect if e.g. a new future was accidentally changing the gas-costs of older evm versions. But I'm not sure whether it would really make sense. I just imagine, if someone adds a feature for the latest evm version, that should only influence the gas cost in that particular version, but it also change the gas cost of other version(s), that this could be interesting to see.

aarlt commented on test/Common.cpp:
```diff
@@ -120,6 +122,11 @@ void CommonOptions::validate() const
+	assertThrow(
+		!enforceGasTest || evmVersion() == langutil::EVMVersion::latest(),
+		ConfigException,
+		"Gas costs can only be enforced on latest evm version."
```
Would it not make sense to support this for all evm versions. E.g. if `EVMVersion` is not defined, it will always run against `latest`. But if we have e.g. `EVMVersion: constantinople` we do the gas measurement exactly for this version, also if we have something like `EVMVersion: >=constantinople`, we could always run against the smallest version defined, e.g. `constantinople` in this case.

aarlt replied:
> > I just checked out your branch and ran the tests locally. It looks like that I have always different gas measurement results.
> 
> It's using `berlin` EVM by default. I'm not sure if that's the default value for `--evm-version` as well. If it's not than that might explain the difference. Can you try to run with `--evm-version berlin`?

maybe the latest version should be default, if not version was not specified.

cameel replied:
It definitely should. And maybe it is. I just have a suspicion that it might be still set to `istanbul`, mostly due to https://github.com/ethereum/solidity/pull/10863#discussion_r566815370.

If latest is not the default, we should change the default it but that would be something for a separate PR in my opinion.

cameel replied:
> But if we have e.g. EVMVersion: constantinople we do the gas measurement exactly for this version, also if we have something like EVMVersion: >=constantinople,

The problem with that is that most tests will probably work only with one specific EVM version. Measurements on other versions will be different and tests will not pass. And we don't really want to put expectations for multiple versions in the isoltest comments (https://gitter.im/ethereum/solidity-dev?at=6012ceb55500a97f82ce5b72).

cameel replied:
Looks like `berlin` is the default after all: https://github.com/ethereum/solidity/blob/v0.8.1/test/Common.cpp#L182-L193

mijovic commented:
@aarlt Regarding what gas costs are added to expectations by enforce feature, I did it in a way that I add only gas costs greater than X(which was set to 100000 gas). So if, for example ir* is missing, it means that it has gas cost of less than 100k gas...

If it is confusing, I'll change it

aarlt commented:
I just checked out your branch and ran the the tests locally. It looks like that I have always different gas measurement results.

If i run `test/tools/isoltest -t semanticTests/\* --vm /Users/alex/evmone/lib/libevmone.dylib  --no-smt` I got this:
```
semanticTests/constructor_ihneritance_init_order_2.sol: FAIL
  Contract:
    contract A {
        uint x = 42;
        function f() public returns(uint256) {
            return x;
        }
    }
    contract B is A {
        uint public y = f();
    }
    // ====
    // compileToEwasm: also
    // compileViaYul: also

  Gas results missing or wrong, obtained result:
  // constructor() ->
  // gas LegacyOptimized: 138540
  // gas Yul: 232112
  // gas YulOptimized: 153851
  // gas Legacy: 151224

  // y() -> 42
  // gas LegacyOptimized: 22092
  // gas Yul: 22744
  // gas YulOptimized: 22262
  // gas Legacy: 22193
```
and in the test file
```
❯ cat ../test/libsolidity/semanticTests/constructor_ihneritance_init_order_2.sol
contract A {
    uint x = 42;
    function f() public returns(uint256) {
        return x;
    }
}
contract B is A {
    uint public y = f();
}
// ====
// compileToEwasm: also
// compileViaYul: also
// ----
// constructor() ->
// gas Legacy: 152088
// gas LegacyOptimized: 138540
// gas Yul: 232112
// gas YulOptimized: 153851
// y() -> 42
// gas Legacy: 22193
// gas LegacyOptimized: 22092
// gas Yul: 22744
// gas YulOptimized: 22262
```

What could be the reason for this?

aarlt commented:
> @aarlt Regarding what gas costs are added to expectations by enforce feature, I did it in a way that I add only gas costs greater than X(which was set to 100000 gas). So if, for example ir* is missing, it means that it has gas cost of less than 100k gas...
> 
> If it is confusing, I'll change it

@mijovic not sure whether it's confusing. If the values less than `100000` changes very often, it may be too annoying to always update the tests. But I guess they would not change that much, right? I think that it would be nice to see always any gas change, so that the implementor will get more conscious what a change meant to the gas costs.

aarlt commented:
I was also thinking whether it could make sense to add an option to the test-case to disable the gas measurements. E.g. something like
```
    contract C {
       ...
    }
    contract D is C {
        constructor(uint newI, uint newK) C(newI, newK + 1) {}
    }
    // ====
    // compileViaYul: also
    // doGasMeasurements: false
    ...
```

cameel commented:
@aarlt 
> I just checked out your branch and ran the tests locally. It looks like that I have always different gas measurement results.

It's using `berlin` EVM by default. I'm not sure if that's the default value for `--evm-version` as well. If it's not than that might explain the difference. Can you try to run with `--evm-version berlin`?

cameel commented:
> I was also thinking whether it could make sense to add an option to the test-case to disable the gas measurements. E.g. something like

That sounds like a good idea if it's not too much effort. I can see how in some cases these gas measurements could be annoying.

cameel commented:
@aarlt 
> I just checked out your branch and ran the tests locally. It looks like that I have always different gas measurement results.

Looking closer at your results, the numbers are actually the same, only the order diffrers.

Maybe the code that prints the `gas` expectations is not enforcing any particular order and it's non-deterministic? If so, we should sort them.

aarlt commented:
> > I just checked out your branch and ran the tests locally. It looks like that I have always different gas measurement results.
> 
> It's using `berlin` EVM by default. I'm not sure if that's the default value for `--evm-version` as well. If it's not than that might explain the difference. Can you try to run with `--evm-version berlin`?

maybe the latest version should be default, if version was not specified.

cameel commented:
@aarlt Looks like it's not hard-coded to any specific version and uses the latest one instead: https://github.com/ethereum/solidity/blob/v0.8.1/test/Common.cpp#L182-L193

So, false alarm, fortunately. The problem is probably just that the order is non-deterministic. Maybe it actually by accidents somehow depends on the evm version object? :)

mijovic commented:
> > @aarlt Regarding what gas costs are added to expectations by enforce feature, I did it in a way that I add only gas costs greater than X(which was set to 100000 gas). So if, for example ir* is missing, it means that it has gas cost of less than 100k gas...
> > If it is confusing, I'll change it
> 
> @mijovic not sure whether it's confusing. If the values less than `100000` changes very often, it may be too annoying to always update the tests. But I guess they would not change that much, right? I think that it would be nice to see always any gas change, so that the implementor/s and reviewer/s will get more conscious what a change meant to the gas costs.

My idea for this `100000` limit was a mechanism to initially activate some tests. Also, if you add particular query to the test it will be measured in any case. So, this enforce gas test was developed for the purpose of activating gas costs.
I would say that we should have option for the test to measure only some gas costs (without need of adding all 4 measures).

mijovic commented:
> @aarlt
> 
> > I just checked out your branch and ran the tests locally. It looks like that I have always different gas measurement results.
> 
> It's using `berlin` EVM by default. I'm not sure if that's the default value for `--evm-version` as well. If it's not then that might explain the difference. Can you try to run with `--evm-version berlin`?

If we don't specify version it runs `istanbul` as that is default one in EVMVersion:
`Version m_version = Version::Istanbul;`

mijovic commented:
> @aarlt Looks like it's not hard-coded to any specific version and uses the latest one instead: https://github.com/ethereum/solidity/blob/v0.8.1/test/Common.cpp#L182-L193
> 
> So, false alarm, fortunately. The problem is probably just that the order is non-deterministic. Maybe it actually by some weird accident somehow depends on the evm version object? :)

Order of elements was not deterministic, but anyway it shouldn't produce failure as order is not important. Now I will make it deterministic anyway.

mijovic committed:
Making order of gas expectations deterministic.

cameel commented:
> Order of elements was not deterministic, but anyway it shouldn't produce failure as order is not important. Now I will make it deterministic anyway.

Making the order deterministic is a good thing. Less noise in diffs. But if it's not that then there must be a difference. I just looked at it more closely and there is one tiny difference after all: @aarlt got `gas Legacy` of 151224 for constructor while the expectation is 152088.

> If we don't specify version it runs `istanbul` as that is default one in EVMVersion:
> `Version m_version = Version::Istanbul;`

Interesting. I missed that detail. I thought that `EVMVersion{}` was defaulting to the latest version. But it makes sense - `berlin` has not been released yet. So this probably explains the difference for @aarlt.

This shows that `isoltest` really should refuse to work and report an error when you use a different EVM. Or at least print a warning when updating expectations.

Also, we should probably use the same value for both? I.e. either run tests by default on latest EVM and use gas costs for do both for current EVM for both?

cameel commented:
I asked about this on gitter and it turns out that we don't have anything implemented for `berlin` and [evmone even uses the same exact set of instructions for `berlin` and `istanbul`](https://github.com/ethereum/evmone/blob/v0.5.0/lib/evmone/instructions.cpp#L1415). So the difference @aarlt reported must have been caused by something else.

And the default should stay at `istanbul`.

mijovic commented:
@aarlt @cameel When I changed version of evm on which we run gas cost some tests have changed :)
I left them as a separate commit, so you can see tham. Will squash it into Update tests commit at the end.

Also, deterministic order have it's own commit so it is easier for you to review

mijovic commented:
> @aarlt @cameel When I changed version of evm on which we run gas cost some tests have changed :)
> I left them as a separate commit, so you can see tham. Will squash it into Update tests commit at the end.
> 
> Also, deterministic order have it's own commit so it is easier for you to review

I ended up squashing commits already.
I think it is ready for another review now.
Also, would be nice to get some more feedback for https://github.com/ethereum/solidity/pull/10863#discussion_r566840010

aarlt commented:
> I asked about this on gitter and it turns out that we don't have anything implemented for `berlin` and [evmone even uses the same exact set of instructions for `berlin` and `istanbul`](https://github.com/ethereum/evmone/blob/v0.5.0/lib/evmone/instructions.cpp#L1415). So the difference @aarlt reported must have been caused by something else.
> 
> And the default should stay at `istanbul`.

The thing was that there was no difference, but a problem in the order. Somehow the expectations where "shuffled", but the values where the same. Not sure what made the difference in the ordering.

cameel commented:
@aarlt Was there really no difference? `gas Legacy` does differ in the output you posted earlier (https://github.com/ethereum/solidity/pull/10863#issuecomment-769964836). I said it's just the order but I was wrong, I didn't look at it carefully enough.

cameel commented on test/Common.cpp:
```diff
@@ -120,6 +122,11 @@ void CommonOptions::validate() const
 		"Invalid test path specified."
 	);
+	assertThrow(
+		!enforceGasTest || evmVersion() == langutil::EVMVersion::istanbul(),
```
I think it should be `EVMVersion{}` here too.

```suggestion
		!enforceGasTest || evmVersion() == langutil::EVMVersion{},
```

cameel commented on test/tools/IsolTestOptions.cpp:
```diff
@@ -76,6 +76,8 @@ bool IsolTestOptions::parse(int _argc, char const* const* _argv)
 		return false;
 	}
 	enforceViaYul = true;
+	enforceGasTest = (evmVersion() == langutil::EVMVersion::istanbul());
```
Again, `EVMVersion{}`:
```suggestion
	enforceGasTest = (evmVersion() == langutil::EVMVersion{});
```

cameel commented on test/tools/IsolTestOptions.cpp:
```diff
@@ -76,6 +76,8 @@ bool IsolTestOptions::parse(int _argc, char const* const* _argv)
 		return false;
 	}
 	enforceViaYul = true;
+	enforceGasTest = (evmVersion() == langutil::EVMVersion::istanbul());
+	enforceGasTestMinValue = 100000;
```
Does this code mean that we're just ignoring the command-line parameters for some options in `isoltest` and just hard-code them to defaults? Why?

If there is a legit reason for this then I still think that just failing and telling the user that his values will be ignored might be better.

mijovic replied:
The thing is that I wanted to enforce gas updates in isoltest, same like for enforcing via yul.
If you think we shouldn't run it, I will disable it...

cameel reviewed (commented):
Looks like almost everything has been addressed. The only remaining things are:
- I think the original suggestion from @chriseth in https://github.com/ethereum/solidity/pull/10863#discussion_r566112827 was to replace `EVMVersion::istanbul()` with `EVMVersion{}`. There was a misunderstanding with `berlin` and also I unnecessarily suggested adding `latest()` (which is confusing if latest is not the default) but I think that now that everything has been cleared up we should go back to that original suggestion.
- Hiding `gas ir` by default requested by @chriseth in https://github.com/ethereum/solidity/pull/10863#discussion_r567842839
- My question about `isoltest` apparently ignoring some command-line options.
- We should probably try to figure out why costs are different with and without `--evm-version istanbul`.
- Tests are not passing yet.

chriseth commented on test/Common.cpp:
```diff
@@ -103,6 +103,8 @@ CommonOptions::CommonOptions(std::string _caption):
 		("no-smt", po::bool_switch(&disableSMT), "disable SMT checker")
 		("optimize", po::bool_switch(&optimize), "enables optimization")
 		("enforce-via-yul", po::bool_switch(&enforceViaYul), "Enforce compiling all tests via yul to see if additional tests can be activated.")
+		("enforce-gas-cost", po::bool_switch(&enforceGasTest), "Enforce gas cost expectations in semantic tests.")
```
Shouldn't this rather be named something like `fail-on-gas-cost-mismatch`? What is this enforcing?

chriseth replied:
Or `check-gas-cosst`?

axic replied:
I thought we said to have something like `--gas-tests-threshold <n>` where n specifies the minimum required gas above which it is included in the test.

mijovic replied:
with enforcing it is going to fail the tests that use more than `X` gas and force us to add this costs to the tests... `X` can be set with the option below.
If you think it is not useful, I will remove it completely.

chriseth commented on test/libsolidity/util/SoltestTypes.h:
```diff
@@ -23,41 +23,42 @@ namespace solidity::frontend::test
+#define SOLT_TOKEN_LIST(T, K)          \
+	T(Unknown, "unknown", 0)           \
+	T(Invalid, "invalid", 0)           \
+	T(EOS, "EOS", 0)                   \
```
Maybe we should just modify the line that requires moving the backslash...

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -271,9 +287,50 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
+		for (auto const& test: m_tests)
+		{
+			ErrorReporter errorReporter;
+			_stream << test.format(errorReporter, _linePrefix, false, _formatted) << endl;
```
What does it print here? Would be nice to see:
 - old value
 - new value
 - absolute change
 - relative change

mijovic replied:
I will modify the output when tests are failing definitely

stackenbotten commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -100,6 +109,9 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 
 	parseExpectations(m_reader.stream());
 	soltestAssert(!m_tests.empty(), "No tests specified in " + _filename);
+	
```
Coding style error

mijovic commented:
I got deterministic results locally. Will wait here to see if they pass in CI as well. If they are passing, will fix remaining TODOs later

mijovic commented:
Added print for obtained and expected, hope it will give some insight why this fails on CI

mijovic committed:
Don't run gas for ir if it is not explicitely added to expectations.

mijovic committed:
Adding debug info on gas costs when updating gas values.

Co-authored-by: Kamil Śliwak <kamil.sliwak@codepoets.it>

mijovic commented:
@cameel  I think it is looking fine to me now. We can do another round of reviews

I would prefer moving deploy gas costs to other PR (in a batch after this one) as this one is already too huge and unreadable...

mijovic commented on test/libsolidity/util/TestFunctionCall.h:
```diff
@@ -42,33 +42,38 @@ namespace solidity::frontend::test
-		bool const _highlight = false
+		bool const _highlight = false,
+		bool const _renderGasCostResult = false,
+		bool const _debugGasPrint = false
```
Need to add docstrimg for this parameter as well

ekpyron commented on test/libsolidity/GasCosts.cpp:
```diff
@@ -91,7 +91,7 @@ BOOST_AUTO_TEST_CASE(string_storage)
 		}
 	)";
-	m_compiler.overwriteReleaseFlag(true);
+	m_compiler.setVersionType(CompilerStack::VersionType::Release);
```
Just as a note: It'd probably make more sense to use ``Empty`` here as well, together with disabling metadata hashes - that'd simplify the macros significantly... but we can do all that separately, no need to do it in this PR.

mijovic replied:
I'll try doing this once we are done with this PR

ekpyron commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -100,6 +109,9 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 	soltestAssert(!m_tests.empty(), "No tests specified in " + _filename);
+
+	m_compiler.setVersionType(CompilerStack::VersionType::Empty);
+	m_compiler.setMetadataHash(CompilerStack::MetadataHash::None);
```
Since we only check gas on the latest EVM version, maybe it's a reasonable compromise to do this only for that evm version? That way at least the full soltest CI runs will run in both combinations. Not that it's extremely likely that it will change test results, but maybe better to still cover it partly that way, if we can?

mijovic replied:
I added this to be set only for default EVM vesion runs now

ekpyron commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -213,6 +228,9 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
 
 			bool outputMismatch = (output != test.call().expectations.rawBytes());
+			if (!outputMismatch && !checkGasCostExpectation(test, _compileViaYul))
+				m_gasCostFailure = true;
```
Maybe also set ``success = false;`` keeping the condition in line 241 simpler?

mijovic replied:
Added, condition should look better now as well.

ekpyron commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -274,6 +306,32 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
 	return TestResult::Success;
 }
 
+bool SemanticTest::checkGasCostExpectation(TestFunctionCall& io_test, bool _compileViaYul) const
```
```suggestion
bool SemanticTest::checkGasCostExpectation(TestFunctionCall& _io_test, bool _compileViaYul) const
```

mijovic replied:
According to https://github.com/ethereum/solidity/blob/develop/CODING_STYLE.md we should use io_test in this case :)
`Exception: "io_parameterName" when it is used for both input and output. See 6(f).`

ekpyron replied:
Haha, ok, fair enough :-). Although there is exactly zero occurrences of those ``o_`` and ``io_`` prefixes in the entire C++ code base :-). Also what's ``6(f)`` :-D?
But yeah, I was actually wondering about it being a weird name in general as well and why it's called "io" - now I see :-).

But yeah, in this case I don't mind either way myself then.

mijovic replied:
Actually, I didn't see this type of names as well. @cameel suggested this change originally :)
I am fine with any, but if we dont stick to io, maybe I can remove it

cameel replied:
Really? I thought that we did have at least at few? If it's a dead rule then we should remove the rule from coding style.

BTW, I think it wouldn't be a bad idea to actually start using a convention for parameters like that. It does make them stand out which is good when reviewing.

ekpyron commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -36,7 +36,9 @@ string TestFunctionCall::format(
 	ErrorReporter& _errorReporter,
 	string const& _linePrefix,
 	bool const _renderResult,
-	bool const _highlight
+	bool const _highlight,
+	bool const _renderGasCostResult,
+	bool const _debugGasPrint
```
This is getting out of hand :-). As seen by the fact that you added ``/* argumentName */`` comments at the call sites...
Can this maybe be reduced to some ``enum`` values instead?
Or an enum and a bool (distinguishing between "debug", i.e. interactive output, and expectation update mode might make sense in general?)

mijovic replied:
Will change to enum as this indeed looks really ugly

mijovic replied:
I tried to make it better, hope it looks good now :)

cameel replied:
I think it's better now, though I also suggested some renaming below. It would be perfect if the last parameter was an enum.

ekpyron commented on test/libsolidity/SemanticTest.h:
```diff
@@ -70,6 +87,10 @@ class SemanticTest: public SolidityExecutionFramework, public EVMVersionRestrict
 	bool m_compileViaYulCanBeSet = false;
+
+	bool m_gasCostFailure = false;
+	bool m_enforceGasCost;
```
```suggestion
	bool m_enforceGasCost = false;
```

cameel replied:
Why though? It always gets initialized by the constructor.

ekpyron commented on test/TestCase.h:
```diff
@@ -41,6 +41,8 @@ class TestCase
 		langutil::EVMVersion evmVersion;
 		std::vector<boost::filesystem::path> vmPaths;
 		bool enforceCompileViaYul;
+		bool enforceGasCost;
```
```suggestion
		bool enforceCompileViaYul = false;
		bool enforceGasCost = false;
```

PODs should just *always* be member-initialized everywhere no matter what, unless there's very good reason not to (which rarely happens)...

ekpyron reviewed (commented):
Some comments, but looking good!

cameel commented on libsolidity/interface/CompilerStack.h:
```diff
@@ -104,6 +104,12 @@ class CompilerStack: boost::noncopyable
+		Release,
+		Prerelease,
+		Empty
+	};
```
`VersionType` sounds pretty generic to me and without looking at the implementation, I probably would not guess what it changes about the compilation. How about changing it to something more self-explanatory?

```suggestion
	enum class MetadataFormat {
		WithReleaseVersionTag,
		WithPrereleaseVersionTag,
		NoMetadata,
	};
```

And also changing `setVersionType()` and `m_versionType` to match.

cameel commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -1596,10 +1599,12 @@ bytes CompilerStack::createCBORMetadata(Contract const& _contract) const
+	else if (m_versionType == VersionType::Prerelease)
 		encoder.pushString("solc", VersionStringStrict);
+	else
+		solAssert(m_versionType == VersionType::Empty, "Invalid version type");
```
There's an `if` at the beginning of the function so the `else` part will never be reached.
```suggestion
	else
	{
		solAssert(m_versionType == VersionType::Prerelease, "Invalid version type");
		encoder.pushString("solc", VersionStringStrict);
	}
```

cameel commented on test/libsolidity/Metadata.cpp:
```diff
@@ -91,7 +91,9 @@ BOOST_AUTO_TEST_CASE(metadata_stamp)
-			compilerStack.overwriteReleaseFlag(release);
+			compilerStack.setVersionType(
+				release ? CompilerStack::VersionType::Release : CompilerStack::VersionType::Prerelease
+			);
```
I think that these tests should be extended to iterate over all three possible values of `VersionType`.

mijovic replied:
https://github.com/ethereum/solidity/issues/10943

cameel replied:
By the way, if you'd like to make this PR smaller then I think that the part related to the metadata flag would actually be a good standalone PR.

cameel commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -319,6 +324,50 @@ string TestFunctionCall::formatRawParameters(
+			differentResults ?
+			(m_gasCosts.at(runType) > m_call.expectations.gasUsed.at(runType)) :
+			false;
+		u256 difference = 0;
+		if (differentResults)
+			difference =
+				increase ?
+				(m_gasCosts.at(runType) - m_call.expectations.gasUsed.at(runType)) :
+				(m_call.expectations.gasUsed.at(runType) - m_gasCosts.at(runType));
```
Does `abs()` work with `boost::multiprecision`? If not, I'd create one and put it somewhere in our common headers.

mijovic replied:
I used it this way as  I wasn't sure about exact conversion it would do as u256 is unsigned.
In any case I should add helper

mijovic replied:
`abs()` works with `boost::multiprecision`, so using that one
Thanks :)

cameel commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -319,6 +324,50 @@ string TestFunctionCall::formatRawParameters(
+			0;
+
+		if (!runType.empty())
+		{
```
I think that this `if` should cover the calculations above too. Why do them if we're not going to use the results?

cameel commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -319,6 +324,50 @@ string TestFunctionCall::formatRawParameters(
+			   	(_debugGasPrint && differentResults && _renderGasCostResult) ?
+				(" ["s + sign + difference.str() + " (" + sign + to_string(percent)) + "%)]" :
+				""s
+			);
```
```suggestion
			if (_debugGasPrint && differentResults && _renderGasCostResult)
				os << " [" << sign << difference << " (" << sign << percent << "%)]";
```

cameel commented on test/libsolidity/util/TestFunctionCall.h:
```diff
@@ -116,6 +126,13 @@ class TestFunctionCall
+		std::string const& _linePrefix,
+		bool const _renderGasCostResult,
+		bool const _debugGasPrint
+	) const;
```
How about these names?
```suggestion
	std::string formatGasExpectations(
		std::string const& _linePrefix,
		bool _useActualCost,
		bool _showDifference
	) const;
```

But it would be much more readable as an enum: `ExpectedCost`, `ActualCost`, `ActualCostAndDiffference`.

Also, I'm not sure if there's much sense in declaring these `bool`s as `const`. I'd personally prefer if they were `const` by default but since they're not, adding `const` manually everywhere seems a bit tedious.

cameel commented on test/libsolidity/SemanticTest.h:
```diff
@@ -70,6 +87,10 @@ class SemanticTest: public SolidityExecutionFramework, public EVMVersionRestrict
 	bool m_compileViaYulCanBeSet = false;
+
+	bool m_gasCostFailure = false;
+	bool m_enforceGasCost = false;
```
`m_enforceGasCost` and `m_enforceViaYul` already have defaults in the constructor.

cameel replied:
I see that @ekpyron suggested adding it in https://github.com/ethereum/solidity/pull/10863/files#r573604188

I'm not really sure why. Constructor already takes care of that and I think that having defaults in too many places just makes it easy for them to go out of sync.

ekpyron replied:
It's **always** better to immediately member-initialize PODs, everywhere, no matter what, no exceptions whatsoever.
If you do, the worst thing that can happen is that you overwrite it by the opposite value in the constructor, but there's really no harm in that. The worst thing that can happen in *not* doing it, is undefined behaviour that will not break for ten years and then all of a sudden make your program explode for no reason ;-).

cameel replied:
Actually the worst that it can happen is that you can forget to actually initialize it with the proper value in constructor and not notice because it won't fail. But yeah, I get your point, in C++ uninitialized stuff will likely not fail outright anyway and might be just as hard to notice. I guess zero-initialization might be safer as a general rule. The biggest issue I had with this was the repeated default value for minimum gas cost but that's gone after the previous review round.

ekpyron replied:
I would say, you can forget to initialize a value in a constructor no matter if you member-initialized it or not, that's the same risk independently. Only in one case what you end up with fails deterministically, in the order randomly :-).
Or put yet in another way: *if* you forget to initialize something in the constructor, but *haven't* member-initialized, your bug is as bad as choosing any possible value for it potentially including the one improbable worst possible value that might really break stuff, so you're up against an entire array of potential misbehaviour - whereas if you do member-initialize, you see and know exactly what will happen and can check how bad it is easily.
So yeah, as you see, I'm an (exaggeratedly ;-)) strong proponent of this ;-).

cameel commented on test/TestCase.h:
```diff
@@ -40,7 +40,9 @@ class TestCase
 		langutil::EVMVersion evmVersion;
 		std::vector<boost::filesystem::path> vmPaths;
-		bool enforceCompileViaYul;
+		bool enforceCompileViaYul = false;
+		bool enforceGasCost = false;
+		u256 enforceGasCostMinValue;
```
Do we really need defaults here? Looks like it used to work fine without them.

cameel replied:
OK, I found an earlier comment by @ekpyron suggesting the opposite change :)

https://github.com/ethereum/solidity/pull/10863#discussion_r570887964

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -100,6 +109,12 @@ SemanticTest::SemanticTest(string const& _filename, langutil::EVMVersion _evmVer
 	parseExpectations(m_reader.stream());
 	soltestAssert(!m_tests.empty(), "No tests specified in " + _filename);
+
+	if (_evmVersion == EVMVersion{})
+	{
+		m_compiler.setVersionType(CompilerStack::VersionType::Empty);
+		m_compiler.setMetadataHash(CompilerStack::MetadataHash::None);
+	}
```
Why only for the latest EVM?

If it's because gas tests only run on the latest EVM then maybe it would be better to make it depend directly on `_enforceGasCost` instead?

cameel replied:
I see that it was suggested by @ekpyron in https://github.com/ethereum/solidity/pull/10863#discussion_r570880001.

So yeah, the change was good but I think that making it depend on `_enforceGasCost` would produce the same effect and be semantically more correct (also, less chance that it will break if we suddenly decide to check gas expectations on more than one version).

cameel commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -141,7 +144,7 @@ string TestFunctionCall::format(
 			bool const isFailure = m_failure;
 			result = isFailure ?
-				formatFailure(_errorReporter, m_call, output, _renderResult, highlight) :
+				formatFailure(_errorReporter, m_call, output, _renderMode == RenderMode::GasResults, highlight) :
```
Is this correct? Before your change this would always pass `true` in the fourth argument when actual results were requested. Now it's false in that case and only for `GasResults` it's `true`.

```suggestion
				formatFailure(_errorReporter, m_call, output, true, highlight) :
```

mijovic replied:
I think It should be `_renderMode == RenderMode::ActualValuesExpectedGas`

cameel commented on test/libsolidity/util/TestFunctionCall.h:
```diff
@@ -42,33 +42,42 @@ namespace solidity::frontend::test
+		Expectations,
+		ResultsOnly,
+		GasResults
+	};
```
How about these? Now very succint but it took me some time before I fully realized what the current values actually mean.
```suggestion
	enum class RenderMode
	{
		ExpectedValuesExpectedGas,
		ActualValuesExpectedGas,
		ExpectedValuesActualGas,
	};
```

Alternatively, you could reduce it to two values (`TestExpectations` and `TestResults`) and instead have two separate parameters, one for gas and one for other values.

cameel commented on test/libsolidity/util/TestFunctionCall.h:
```diff
@@ -42,33 +42,42 @@ namespace solidity::frontend::test
+	/// if it is Expectations we show test expectations, and in case of GasResults we print test expectations along with actual gas costs
 	/// If _highlight is false, it's formatted without colorized highlighting. If it's true, AnsiColorized is
 	/// used to apply a colorized highlighting.
+	/// If _interactivePrint is true, we are running interactive print about failure
```
```suggestion
	/// If _interactivePrint is true, we are producing output that will be interactively shown to the user in the terminal rather than used to update the expectations in the test file.
```

cameel commented on test/libsolidity/util/TestFunctionCall.h:
```diff
@@ -42,33 +42,42 @@ namespace solidity::frontend::test
 	/// Formats this function call test and applies the format that was detected during parsing.
-	/// If _renderResult is false, the expected result of the call will be used, if it's true
-	/// the actual result is used.
+	/// _renderMode render mode that we are currently executing. If it is ResultsOnly we show actual results,
+	/// if it is Expectations we show test expectations, and in case of GasResults we print test expectations along with actual gas costs
```
How about this?

```suggestion
	/// _renderMode determines the source of values to be inserted into the updated test expectations.
	///     RenderMode::ResultsOnly: use the values produced by the test but for gas preserve the original expectations,
	///     RenderMode::Expectations: preserve the original expectations for both gas and other values,
	///     RenderMode::GasResults: use the original expectations but for gas use actual values,
```

mijovic committed:
Setting metadata has and version type for semantic tests to be empty.

mijovic committed:
Changing enforceGasCost functionality.

mijovic committed:
Ading CI job to check gas costs.

mijovic committed:
Update tests.

mijovic commented:
@cameel I think it is ready for review again

mijovic commented:
Also, I changed to run gas tests only if `--enforce-gas-cost` is passed to `soltest`, and by default in `isoltest` it runs them.
Because of that, I added a CI jobs for optimized and unoptimized gas cost checks.
Then I changed occurrences of `if (m_evmVersion== EVMVersion{}` to `if (m_enforceGasCost)`

Also, did change a bit debug print by using signed operations, so no need to complicate with `increase` flag.

cameel commented on .circleci/soltest_all.sh:
```diff
@@ -65,6 +65,14 @@ STEP=$(($STEP + 1))
 [[ " $RUN_STEPS " =~ " $STEP " ]] && EVM=istanbul OPTIMIZE=1 ABI_ENCODER_V1=1 BOOST_TEST_ARGS="-t !smtCheckerTests" "${REPODIR}/.circleci/soltest.sh"
 STEP=$(($STEP + 1))
 
+# Run gas tests on semanticTests without optimizer.
+[[ " $RUN_STEPS " =~ " $STEP " ]] && EVM=istanbul OPTIMIZE=0 ABI_ENCODER_V1=0 BOOST_TEST_ARGS="-t semanticTests/*" SOLTEST_FLAGS="--enforce-gas-cost" "${REPODIR}/.circleci/soltest.sh"
+STEP=$(($STEP + 1))
+
+# Run gas tests on semanticTests with optimizer.
+[[ " $RUN_STEPS " =~ " $STEP " ]] && EVM=istanbul OPTIMIZE=1 ABI_ENCODER_V1=0 BOOST_TEST_ARGS="-t semanticTests/*" SOLTEST_FLAGS="--enforce-gas-cost" "${REPODIR}/.circleci/soltest.sh"
+STEP=$(($STEP + 1))
+
```
You could put this inside the outer loop below.

mijovic replied:
I added in loop, and actually was no need to add additional jobs.

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -244,15 +268,27 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
+					TestFunctionCall::RenderMode::ExpectedValuesExpectedGas,
+					_formatted,
+					/* _interactivePrint */ true
+				) << endl;
```
```suggestion
			_stream << test.format(
				errorReporter,
				_linePrefix,
				TestFunctionCall::RenderMode::ExpectedValuesExpectedGas,
				_formatted,
				/* _interactivePrint */ true
			) << endl;
```

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -244,15 +268,27 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
+					m_gasCostFailure ? TestFunctionCall::RenderMode::ExpectedValuesActualGas : TestFunctionCall::RenderMode::ActualValuesExpectedGas,
+					_formatted,
+					/* _interactivePrint */ true
+				) << endl;
```
```suggestion
			_stream << test.format(
				errorReporter,
				_linePrefix,
				m_gasCostFailure ? TestFunctionCall::RenderMode::ExpectedValuesActualGas : TestFunctionCall::RenderMode::ActualValuesExpectedGas,
				_formatted,
				/* _interactivePrint */ true
			) << endl;
```

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -244,15 +268,27 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
+			_stream << test.format(
+					errorReporter,
+					_linePrefix,
+					m_gasCostFailure ? TestFunctionCall::RenderMode::ExpectedValuesActualGas : TestFunctionCall::RenderMode::ActualValuesExpectedGas,
```
I think that something like `ActualValuesActualGas` mode would be more appropriate here.

If I understand this correctly, you use `ExpectedValuesActualGas` because actual values should match expected ones in that case anyway (because gas checks are skipped if they're not)? But using actual values would be safer because this assumption might change in the future. E.g. we could add some tolerances in comparisons and have checks pass even if values are not identical.

```suggestion
					m_gasCostFailure ? TestFunctionCall::RenderMode::ActualValuesActualGas : TestFunctionCall::RenderMode::ActualValuesExpectedGas,
```

mijovic replied:
I was thinking to do like this, but there were some tests that were passing checks, but when updating it updates the test when it really shouldn't. Mostly because of encoding. I would prefer to finish up this PR without tackling this one now, in order to have gas costs as soon as possible, then would add this change in separate PR if that is fine with you?

cameel replied:
Yeah, sounds fine. It might be a potential future problem but the code works as it is now so it's not really a high priority.

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -274,6 +310,32 @@ TestCase::TestResult SemanticTest::runTest(ostream& _stream, string const& _line
+	// or setting is "ir" and it's not included in expectations
+	if (
+		!m_enforceGasCost ||
+		solidity::test::CommonOptions::get().useABIEncoderV1 ||
```
`CommonOptions::get().useABIEncoderV1` depends only on the command-line option (and not on the pragma in the file), right?

Then I think you should add that case to `CommonOptions::validate()` and just put an assertion here.

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -321,7 +383,11 @@ void SemanticTest::printSource(ostream& _stream, string const& _linePrefix, bool
-		_stream << test.format("", true, false) << endl;
+		_stream << test.format(
+			"",
+			m_gasCostFailure ? TestFunctionCall::RenderMode::ExpectedValuesActualGas : TestFunctionCall::RenderMode::ActualValuesExpectedGas,
```
I'd use `ActualValuesActualGas` here too:

```suggestion
			m_gasCostFailure ? TestFunctionCall::RenderMode::ActualValuesActualGas : TestFunctionCall::RenderMode::ActualValuesExpectedGas,
```

cameel replied:
Marking as resolved. See https://github.com/ethereum/solidity/pull/10863#discussion_r578517480

cameel commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -319,6 +324,40 @@ string TestFunctionCall::formatRawParameters(
+{
+	stringstream os;
+	for (auto const& [runType, gasUsed]: (_useActualCost ? m_gasCosts : m_call.expectations.gasUsed))
+	{
```
```suggestion
```

cameel commented on test/libsolidity/util/TestFunctionCall.cpp:
```diff
@@ -124,11 +127,11 @@ string TestFunctionCall::format(
 			bool const isFailure = m_call.expectations.failure;
 			result = isFailure ?
-				formatFailure(_errorReporter, m_call, m_rawBytes, _renderResult, highlight) :
+				formatFailure(_errorReporter, m_call, m_rawBytes, _renderMode == RenderMode::ActualValuesExpectedGas, highlight) :
```
This always evaluates to `false` in this branch of the loop:
```suggestion
				formatFailure(_errorReporter, m_call, m_rawBytes, /* _renderResult */ false, highlight) :
```

cameel reviewed (commented):
I have some suggestions but they're mostly cosmetic. Well, maybe except for `ActualValuesActualGas` but this one applies to a hypothetical situation anyway.

So I think that this PR is finished functionality-wise if there is a consensus that we do want this feature.

@mijovic You can ping me when done with the changes. I'll try to review an accept as soon as I see it so that we can finally get done with this PR :)

ekpyron commented:
What's the state of this? This has been pretty much ready for quite a while hasn't it? And it would be very nice to have, so why don't we finish it entirely and merge it :-)?

chriseth commented on .circleci/soltest_all.sh:
```diff
@@ -65,6 +65,14 @@ STEP=$(($STEP + 1))
 [[ " $RUN_STEPS " =~ " $STEP " ]] && EVM=istanbul OPTIMIZE=1 ABI_ENCODER_V1=1 BOOST_TEST_ARGS="-t !smtCheckerTests" "${REPODIR}/.circleci/soltest.sh"
 STEP=$(($STEP + 1))
 
+# Run gas tests on semanticTests without optimizer.
```
Why do we need two additional runs? Can't we just enforce the gas costs when we are running this combination of settings anyway?

mijovic replied:
I moved it now in a loop below

chriseth commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -40,12 +40,21 @@ using namespace boost::unit_test;
+	string const& _filename,
+	langutil::EVMVersion _evmVersion,
+	vector<boost::filesystem::path> const& _vmPaths,
+	bool _enforceViaYul,
```
We could think about moving those into a setting struct.

mijovic replied:
Can we do it in separate PR, this one is already huge? I can do it as a follow up to this one.

chriseth replied:
Sure!

chriseth commented on test/libsolidity/semanticTests/viaYul/array_function_pointers.sol:
```diff
@@ -24,6 +24,18 @@ contract C {
 // compileViaYul: also
 // ----
 // f(uint256,uint256): 1823621, 12323 -> FAILURE
+// gas irOptimized: 100000000
```
Is this because of failure?

mijovic replied:
I guess it is because of failure, but not entirely sure

chriseth replied:
Can you look into it, please?

cameel replied:
It looks to me like it might be just running out of gas trying to allocate a memory array with 1823621 elements rather than failing with a panic.

cameel replied:
By the way, I see that some other tests have a comment like `FAILURE # Out-of-gas #` to make it apparent that this is the expectation. Adding one might be useful here too.

bshastry replied:
Would be nice to print `OUTOFGAS` if `gas == maxGasAllotted`

mijovic replied:
I'll check if it is out of gas, and then will add comment here as well.

mijovic replied:
EVM returns out of gas, so I added comment to make the reason of failure explicit.

mijovic committed:
[isoltest] Adding gas used as semantic tests expectation.

mijovic committed:
Show both expected and obtained gas expectations when failure happens.

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -300,6 +336,31 @@ TestCase::TestResult SemanticTest::runTest(
+		(m_optimiserSettings == OptimiserSettings::full() ? "Optimized" : "");
+
+	// We don't check gas if enforce gas cost is not active
+	// or in case we run abiencoderv1
```
I'd add an assertion about ABI encoder here. It's good that it's a parameter check now but, well, it would also be good to have a sanity check here :)

mijovic replied:
Added assertion

cameel reviewed (dismissed):
I suggested adding an assert in one place and there's a question about some tests running out of gas from @chriseth but overall the PR looks ready to merge to me.

mijovic commented:
@cameel Updated :)

mijovic committed:
Not showing gas used if it is equal to m_gas(in case of out of gas failures).

axic commented on test/libsolidity/semanticTests/inheritance/inherited_function_calldata_calldata_interface.sol:
```diff
@@ -25,3 +25,5 @@ contract B {
 // ----
 // g() -> 42
+// gas irOptimized: 107179
+// gas legacy: 117797
```
Why are some contracts missing `legacyOptimized` or `irOptimized`?

mijovic replied:
If they are less then 100K gas, as that was original threshold for including gas expectation

axic replied:
Ah, this is rather confusing then. Ideally if any of the runs hit the threshold then all of them should be included in the file. But due to the implementation this may be hard to achieve.

mijovic replied:
I think it is possible to add that upgrade with not too much of effort. If you think it is useful, I'll implement it in that way.

mijovic replied:
For optimized it is easier to add both in case that one went over 100k, as both of them run in same setup of soltest.
maybe it is worth trying maling this better
